lte/gateway/c/oai/lib/pipelined_client/PipelinedServiceClient.cpp

```diff
@@ -0,0 +1,713 @@
+
+  local_response->set_response_reader(std::move(response_reader));
+
+  return RETURNok;
+}
+
+//------------------- FORWARDING DATA on TUNNEL -------------------
+//                    FORWARD : v4
+//-------------------------------------------------------------------------
+int PipelinedServiceClient::UpdateUEIPv4SessionSet(
+    const struct in_addr& ue_ipv4_addr, uint32_t in_teid,
+    uint32_t flow_precedence, uint32_t ue_state,
+    std::function<void(grpc::Status, UESessionContextResponse)> callback) {
+  UESessionSet request = UESessionSet();
+  UESessionContextResponse response;
+
+  PipelinedServiceClient& client = get_instance();
+
+  // Set the UE IPv4 address
+  set_ue_ipv4_addr(ue_ipv4_addr, request);
+
+  // Set the incoming and outgoing teid
+  request.set_in_teid(in_teid);
+
+  // Set the precedence
+  request.set_precedence(flow_precedence);
+
+  // Set the ue_state
+  config_ue_session_state(ue_state, request);
+
+  auto local_response = new AsyncLocalResponse<UESessionContextResponse>(
+      std::move(callback), RESPONSE_TIMEOUT);
+
+  auto response_reader = client.stub_->AsyncUpdateUEState(
+      local_response->get_context(), request, &client.queue_);
+  local_response->set_response_reader(std::move(response_reader));
+  return RETURNok;
+}
+
+//                    FORWARD : v4 with flow_dl
+//-------------------------------------------------------------------------
+int PipelinedServiceClient::UpdateUEIPv4SessionSetWithFlowdl(
+    const struct in_addr& ue_ipv4_addr, uint32_t in_teid,
+    const struct ip_flow_dl& flow_dl, uint32_t flow_precedence,
+    uint32_t ue_state,
+    std::function<void(Status, UESessionContextResponse)> callback) {
+  UESessionSet request = UESessionSet();
+  UESessionContextResponse response;
+
+  PipelinedServiceClient& client = get_instance();
+
+  // Set the UE IPv4 address
+  set_ue_ipv4_addr(ue_ipv4_addr, request);
+
+  // Set the incoming and outgoing teid
+  request.set_in_teid(in_teid);
+
+  // Flow dl
+  request.mutable_ip_flow_dl()->CopyFrom(to_proto_ip_flow_dl(flow_dl));
+
+  // Set the precedence
+  request.set_precedence(flow_precedence);
+
+  // Set the ue_state
+  config_ue_session_state(ue_state, request);
+
+  auto local_response = new AsyncLocalResponse<UESessionContextResponse>(
+      std::move(callback), RESPONSE_TIMEOUT);
+
+  auto response_reader = client.stub_->AsyncUpdateUEState(
+      local_response->get_context(), request, &client.queue_);
+  local_response->set_response_reader(std::move(response_reader));
+  return RETURNok;
+}
+
+//                    FORWARD : v4v6
+//-------------------------------------------------------------------------
+int PipelinedServiceClient::UpdateUEIPv4v6SessionSet(
+    const struct in_addr& ue_ipv4_addr, struct in6_addr& ue_ipv6_addr,
+    uint32_t in_teid, uint32_t flow_precedence, uint32_t ue_state,
+    std::function<void(grpc::Status, UESessionContextResponse)> callback) {
+  UESessionSet request = UESessionSet();
```

Comment:
Same comment as above about moving the conversion logic into the other file.

Reply:
Comments addressed.